### DISCUSSION TITLE
[Vertx] How to use a dedicated worker pool

### DISCUSSION THREAD

**Discussion by b-thiswatch** [quarkusio/quarkus#41695] [Q&A]
Hi,

can anyone jump in and help me understand how to use a dedicated vertx pool, when consuming events? I have tried so many variations, but I simply don't get how it's supposed to be done.

I have a process that creates several thumbnails for images, stores them and creates a database entry.
Below are two ways that somehow looked promising, but they only utilise 1 thread of the thumbs worker.

```kotlin
@ApplicationScoped
class EventHandler(
    private val ensureThumbnails: EnsureThumbnails,
    vertx: Vertx,
) : Loggable {
    private val thumbsPool: WorkerExecutor =
        vertx.createSharedWorkerExecutor("thumbs", 4, 5, TimeUnit.MINUTES)

    @ConsumeEvent(FileEvents.ENSURE_THUMBS)
    fun process(cmd: EnsureThumbsCommand) {
        thumbsPool
            .executeBlockingAndForget {
                val ctx = Arc.container().requestContext()
                ctx.activate()
                ensureThumbnails.ensureThumbnails(cmd.fileReferenceId, cmd.wantedSizes)
                ctx.terminate()
            }
    }
}
```

Here I did not use `@Blocking` on the consume event, but then I had to activate the context myself.

```kotlin
@ApplicationScoped
class EventHandler(
    private val ensureThumbnails: EnsureThumbnails,
    vertx: Vertx,
) : Loggable {
    private val thumbsPool: WorkerExecutor =
        vertx.createSharedWorkerExecutor("thumbs", 4, 5, TimeUnit.MINUTES)

    @ConsumeEvent(FileEvents.ENSURE_THUMBS)
    @Blocking
    fun process(cmd: EnsureThumbsCommand) {
        thumbsPool
            .executeBlocking { 
                ensureThumbnails.ensureThumbnails(cmd.fileReferenceId, cmd.wantedSizes)
            }.await().indefinitely()
    }
}
```

This one looks a bit better for my taste, but also only utilises one thread of the thumbs pool.

Any help/advise is appreciated

Thanks

edit: Would something like this be a welcomed feature?

```kotlin
@ApplicationScoped
class EventHandler(
    private val ensureThumbnails: EnsureThumbnails,
    vertx: Vertx,
) : Loggable {
    private val thumbsPool: WorkerExecutor =
        vertx.createSharedWorkerExecutor("thumbs", 4, 5, TimeUnit.MINUTES)

    @ConsumeEvent(FileEvents.ENSURE_THUMBS)
    @WorkExecutor(name = "thumbs") // <-- you get the idea...
    @Blocking
    fun process(cmd: EnsureThumbsCommand) = ensureThumbnails.ensureThumbnails(cmd.fileReferenceId, cmd.wantedSizes)

}
```

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
cc @cescoffier

**Comment by mschorsch** [comment]
You could try using a ManagedExecutor (https://quarkus.io/guides/context-propagation#usage-example-for-completionstage) instead of the Vert.x worker pool. The ManagedExecutor propagates the context.

**Reply by b-thiswatch** [reply]
Thanks, will have a look at this. Skimming through the docs, I am not sure if this solves my problem, though. The docs say it uses the Quarkus Thread Pool.

My problem is this: I have 250k items to process. This runs as an event based process through the vertx eventbus. On a 10-core machine it takes about 8 mins with which I am quite satisfied. Those 250k items are linked to ~450k images. During processing of the items, I want to publish events to generate thumbs for those images. Generating those thumbs (5 different sizes, webp) takes quite some time. Since I need a more generous timeout and don't want it to interfere with the other process, I thought to use a dedicated thumbs worker pool. The generation of thumbs can also be triggered at any time as a separate process, where it wouldn't matter, if it's a dedicated pool or not, but I like the idea of having those long-running tasks separated.

**Reply by cescoffier** [reply]
The simplest with Java 21 is to use `@RunOnVirtualThread`.
If you want to use your own executor, the simplest is to dispatch the work to that worker thread. If there is a "result" to produce, do something like:

```java
@ConsumeEvent("...")
public Uni<String> process(String input) {
    return Uni.createFrom().item(input)
         .emitOn(myExecutor)
         .map(i -> doTheWork(i)); // Will be called on a thread of the executor        
}

**Reply by b-thiswatch** [reply]
@cescoffier thanks! Will check out the emitOn
I have tried `@RunOnVirtualThread` but end up with only `Unable to acquire JDBC Connection` errors.

**Reply by cescoffier** [reply]
ah, that's interesting. How much concurrency do you have?

**Reply by b-thiswatch** [reply]
I'm pushing ~450k events and the system is running on 10 cores. But the `Unable to acquire JDBC Connection` happens on the first execution already.

Or did I misunderstand?

**Reply by b-thiswatch** [reply]
Sorry, no, that was wrong. It does not fail immediately, but rather really quickly. I will try to capture

**Comment by b-thiswatch** [comment]
Maybe, I should rephrase the question. What is the easiest way to add and use a second worker pool (with a different number of workers and different timeouts) to a Quarkus App?

**Reply by cescoffier** [reply]
You can just define your own scheduler, but be aware that context propagation may not be supported.
